worlds/lingo/static_logic.py

```diff
@@ -0,0 +1,410 @@
+            if rwnd:
+                self.REQUIRED_PAINTING_WHEN_NO_DOORS_ROOMS.append(room_name)
+        else:
+            rwnd = False
+
+        if "exit_only" in painting_data:
+            exit_only = painting_data["exit_only"]
+        else:
+            exit_only = False
+
+        if "enter_only" in painting_data:
+            enter_only = painting_data["enter_only"]
+        else:
+            enter_only = False
+
+        required_door = None
+        if "required_door" in painting_data:
+            door = painting_data["required_door"]
+            required_door = RoomAndDoor(
+                door["room"] if "room" in door else room_name,
+                door["door"]
+            )
+
+        painting_obj = Painting(painting_id, room_name, enter_only, exit_only, orientation,
+                                required_painting, rwnd, required_door, disable_painting, move_painting)
+        self.PAINTINGS[painting_id] = painting_obj
+        self.PAINTINGS_BY_ROOM[room_name].append(painting_obj)
+
+    def process_progression(self, room_name, progression_name, progression_doors):
+        self.PROGRESSIVE_ITEMS.append(progression_name)
+
+        progression_index = 1
+        for door in progression_doors:
+            if isinstance(door, Dict):
+                door_room = door["room"]
+                door_door = door["door"]
+            else:
+                door_room = room_name
+                door_door = door
+
+            room_progressions = self.PROGRESSION_BY_ROOM.setdefault(door_room, {})
+            room_progressions[door_door] = Progression(progression_name, progression_index)
+            progression_index += 1
+
+    def process_room(self, room_name, room_data):
+        room_obj = Room(room_name, [])
+
+        if "entrances" in room_data:
+            for source_room, doors in room_data["entrances"].items():
+                self.process_entrance(source_room, doors, room_obj)
+
+        if "panels" in room_data:
+            self.PANELS_BY_ROOM[room_name] = dict()
+
+            for panel_name, panel_data in room_data["panels"].items():
+                self.process_panel(room_name, panel_name, panel_data)
+
+        if "doors" in room_data:
+            self.DOORS_BY_ROOM[room_name] = dict()
+
+            for door_name, door_data in room_data["doors"].items():
+                self.process_door(room_name, door_name, door_data)
+
+        if "paintings" in room_data:
+            self.PAINTINGS_BY_ROOM[room_name] = []
+
+            for painting_data in room_data["paintings"]:
+                self.process_painting(room_name, painting_data)
+
+        if "progression" in room_data:
+            for progression_name, progression_doors in room_data["progression"].items():
+                self.process_progression(room_name, progression_name, progression_doors)
+
+        self.ROOMS[room_name] = room_obj
+        self.ALL_ROOMS.append(room_obj)
+
+    def __init__(self):
```

Comment:
Should put `__init__` at the top of the class.

In general here, I understand the intention of building up a "static" object, but I feel like the way it's created here is misleading and has some potential issues. You're defining a class object which has a bunch of attributes like `ROOMS` and `PANELS` _on_ the class object itself. And then creating an instance of that class object when you define your world class. Creating that instance loads the yaml file and triggers a bunch of these `process` functions to put the data in the yaml into the class. I'm actually not sure whether the data you're filling ends up on the instance or on the class. I suppose it might depend on whether you ever replace the objects or just modify them.

For instance, I see `self.ROOMS[room_name] = room_obj` above. I believe that's modifying the class object's copy of `ROOMS`. But doing so through `self` is misleading because `self` is an instance. If you were to ever do `self.ROOMS = ...`, that would create a `ROOMS` on the instance and future modifications to `ROOMS` would not touch the class.

Run this snippet to see what I'm talking about in action:

```python
class MyClass:
  a = {}
  b = {}
  
  def __init__(self):
    self.a["c"] = 10
  
    self.b = {}
    self.b["d"] = 10

instance = MyClass()

print(instance.a)  # {'c': 10}
print(instance.b)  # {'d': 10}
print(MyClass.a)   # {'c': 10}
print(MyClass.b)   # {}
```

All this basically just to say why not just write a function that creates an object with these attributes and return that? Then it will be unambiguous that your world object is holding an object with data instead of a class instance that reaches into the class object. If you want immutable data, use immutable data types. Even this current solution is still mutable.

Reply:
Ah, I see there were already some similar comments for the other similar classes.